### ISSUE TITLE
Preserving order of ctor args

### ISSUE THREAD

**Issue by ahjohannessen** [spray/spray-json#119]
Hi,

When I upgraded from spray-json 1.2.6 to 1.3.0 I noticed a change with regards to generated json. The below test did not fail earlier. 

Is ordering of fields in generated json for a case class not consistent with its constructor args?

``` scala
...
case class Bibby(
  id: Int, firstName: String, lastName: String, cid: UUID, time: ZonedDateTime
)
...

"Ordering" should "be as constructor args" in {

  val body =
    HttpEntity(ContentType(`application/json`, `UTF-8`),
    """{
      |  "id": 1,
      |  "firstName": "Jack",
      |  "lastName": "Black",
      |  "cid": "4adbc012-617b-487b-ab31-ac8ba7892714",
      |  "time": "2014-05-23T10:30:00Z"
      |}""".
      stripMarginWithNewline("\n")
    )

  val bibby = Bibby(
    id = 1,
    firstName = "Jack",
    lastName = "Black",
    cid = UUID.fromString("4adbc012-617b-487b-ab31-ac8ba7892714"),
    time = ZonedDateTime.parse("2014-05-23T10:30:00Z")
  )

  marshal(bibby) should equal(Right(body))
}
```

This results in a failing test:

``` scala
[info] - should be as constructor args *** FAILED ***
[info]   Right(HttpEntity(application/json; charset=UTF-8,{
[info]     "lastName": "Black",
[info]     "firstName": "Jack",
[info]     "id": 1,
[info]     "cid": "4adbc012-617b-487b-ab31-ac8ba7892714",
[info]     "time": "2014-05-23T10:30:00Z"
[info]   })) did not equal Right(HttpEntity(application/json; charset=UTF-8,{
[info]     "id": 1,
[info]     "firstName": "Jack",
[info]     "lastName": "Black",
[info]     "cid": "4adbc012-617b-487b-ab31-ac8ba7892714",
[info]     "time": "2014-05-23T10:30:00Z"
[info]   })) ...
```

Just to note, there is nothing wrong with the (un)marshalling and the following passes:

``` scala
marshal(bibby).fold(
  t ⇒ fail(t),
  e ⇒ e.as[Bibby] should equal(Right(bibby))
)
```

**Comment by jeffrey-aguilera** [comment]
print composed with parse should be the identity function.  Changing to a ListMap breaks this most fundamental expectation.

**Comment by betandr** [comment]
Clearly the performance improvement is really important although there are some situations where we really do want some JSON to be in an 'arbitrary' order; would there be any way we can actually specify a render order for this? It'd be really useful.

**Comment by sirthias** [comment]
I agree with @betandr here. JSON members are unordered, so if your code relies on any ordering you should establish it yourself rather than requiring your dependencies to adhere to any particular order.

**Comment by tohenryliu** [comment]
I do a agree with the points made here. Json fields ordering is not in the spec. 
Though a config option possibly in those JsonFormats wouldn't hurt. For applications where performance isn't critical, order preserving can be had with a trade-off. 
Also considering the code needed to expose that is not a lot. 
Or is it still too much a distraction from the library's principal?

**Comment by sirthias** [comment]
@tohenryliu Can't you satisfy your ordering needs on your layer? Similarly to what @betandr described?

**Comment by tohenryliu** [comment]
I think my request is a bit unreasonable. We do not need ordering for any machine consumers. 
It's only useful when I exam the json manually, possibly try to match with the case class values for dev/test purposes. 
also when diff between different results. 

I don't feel so strongly, but the team has developed the habit to rely on that. 

would you accept a Pull Request if I attempt to wrote something for that?

**Comment by sirthias** [comment]
@tohenryliu If you'd like to establish a specific kind of structure for your raw JSON (e.g. for your tests) you could always add a separate reformatting stage that does exactly what you want.
For example, some people might prefer the "original" member order as in the respective domain model (case) class and others an alphabetical order.

**Comment by betandr** [comment]
Although maybe we should question tests that rely on spring-json anyway. Maybe it'd be best to mock objects or have static JSON fixtures rather than testing that spray-json returns the correct value. Testing your own code which uses spray-json may well be the wrong thing.

The really good thing about stuff like this is that it often exposes bad practices and makes us think about how we're approaching things. :)

**Comment by maspwr** [comment]
Might as well throw my updated opinion into the mix. Thinking of this more, I'm quite happy with spray-json doing what it does well, and keeping a small API. I think the most I'd request at this point is an updated docs section mentioning this behavior and possibly even an example how one could enforce ordering on their own.

**Comment by fommil** [comment]
There is no ordering guarantee in the JSON format (or indeed in the containing JavaScript).

I've often thought about writing a custom serialiser for sorted maps and the like, but ultimately preservation of ordering in Spray JSON would be futile because as soon as it goes through any other service it could change. Assuming ordering in maps would mean that you're restricting your use of JSON to "only things that produce and consume it with the ordering caveat".

IMHO, if one requires order in JSON then the only way to enforce it is to use an array representation.

**Comment by magicgoose** [comment]
I'd +1 this, the option to keep declared order is a must for a library like this.

> preservation of ordering in Spray JSON would be futile because as soon as it goes through any other service it could change

That's a totally different problem and it's the user's problem. There _are_ use cases for preserving declaration order. Why not have it? I'm sure it's not too hard :)

**Comment by betandr** [comment]
Quote @magicgoose 
> That's a totally different problem and it's the user's problem. 

If JSON is unmarshalled to objects and and marshalled again into JSON then it could very well be in a different order and while it's one argument to say that it's the user's problem, it's really a non-deterministic problem and relying on order of appearance is a Bad Idea(tm).

> There are use cases for preserving declaration order. Why not have it? 

The JSON spec explicitly says:

"An object is an unordered set of name/value pairs"

If you do have a collection of objects, putting them in an array will preserve the order, however.

> I'm sure it's not too hard :)

Famous last words! ;) Although the reason the order is not preserved, as @sirthias has previously explained, is because the implementation being used to handle JSON would be much slower if it had to do this. We really should be doing everything we can to dissuade people from a really bad practice of relying on insertion order...there's always an alternative!
